hack/tools/logcheck/main.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"go/ast"
+
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/singlechecker"
+)
+
+// Doc explaining the tool.
+const Doc = "Tool to check use of unstructured logging patterns."
+
+// Analyzer runs static analysis.
+var Analyzer = &analysis.Analyzer{
+	Name: "logcheck",
+	Doc:  Doc,
+	Run:  run,
+}
+
+func main() {
+	singlechecker.Main(Analyzer)
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+
+	for _, file := range pass.Files {
+
+		ast.Inspect(file, func(n ast.Node) bool {
+
+			// We are intrested in function calls, as we want to detect klog.* calls
+			// passing all function calls to checkForFunctionExpr
+			if fexpr, ok := n.(*ast.CallExpr); ok {
+
+				checkForFunctionExpr(fexpr.Fun, pass)
+			}
+
+			return true
+		})
+	}
+	return nil, nil
+}
+
+// checkForFunctionExpr checks for unstructured logging function, prints error if found any.
+func checkForFunctionExpr(fun ast.Expr, pass *analysis.Pass) {
+
+	/* we are extracting external package function calls e.g. klog.Infof fmt.Printf
+	   and eliminating calls like setLocalHost()
+	   basically function calls that has selector expression like .
+	*/
+	if selExpr, ok := fun.(*ast.SelectorExpr); ok {
+		// extracting function Name like Infof
+		fName := selExpr.Sel.Name
+
+		// for nested function cases klog.V(1).Infof scenerios
+		// if selExpr.X contains one more caller expression which is selector expression
+		// we are extracting klog and discarding V(1)
+		if n, ok := selExpr.X.(*ast.CallExpr); ok {
+			if _, ok = n.Fun.(*ast.SelectorExpr); ok {
+				selExpr = n.Fun.(*ast.SelectorExpr)
+			}
+		}
+
+		// extracting package name
+		pName, ok := selExpr.X.(*ast.Ident)
```

Comment:
This is taking a shortcut. Was that intentional because the alternative is too complex?

What we want to compare against is `k8s.io/klog/v2`, not just the name that the package was imported under. As it stands now, the following code will not be recognized as using klog.Info:
```
import glog "k8s.io/klog/v2"

func someFunc() {
  glog.Info("hello world"
}
```

It must be possible to lookup the selector, but after staring at the `analysis.Pass`  for a while I still haven't figured out how. Any ideas?

Reply:
I found a solution. This here works and also covers other cases that weren't handled before, like `if klogV := klog.V(1); klogV.Enabled() { klogV.Info("should not be used") }`:
```
// isKlog checks whether an expression is klog.Verbose or the klog package itself.
func isKlog(expr ast.Expr, pass *analysis.Pass) bool {
	// For klog.V(1) and klogV := klog.V(1) we can decide based on the type.
	if typeAndValue, ok := pass.TypesInfo.Types[expr]; ok {
		switch t := typeAndValue.Type.(type) {
		case *types.Named:
			typeName := t.Obj()
			name := typeName.Name()
			pkg := typeName.Pkg()
			pkgPath := pkg.Path()
			if name == "Verbose" && pkgPath == "k8s.io/klog/v2" {
				return true
			}
		}
	}

	// In "klog.Info", "klog" is a package identifier. It doesn't need to
	// be "klog" because here we look up the actual package.
	if ident, ok := expr.(*ast.Ident); ok {
		if object, ok := pass.TypesInfo.Uses[ident]; ok {
			switch object := object.(type) {
			case *types.PkgName:
				pkg := object.Imported()
				if pkg.Path() == "k8s.io/klog/v2" {
					return true
				}
			}
		}
	}

	return false
}
```

I'm currently collecting some improvements of logcheck for contextual logging and will include this in that patch set.

IMHO it's not important enough to be fixed right away. Import renaming was done in some projects that migrated from glog, but seems less common nowadays.
```